data_gateway/data_gateway.py

```diff
@@ -142,28 +147,37 @@ def start(self, stop_when_no_more_data_after=False):
         reader_process.start()
         parser_process.start()
 
-        if self.interactive:
-            interactive_commands_thread = threading.Thread(
-                name="InteractiveCommandsThread",
-                target=self._send_commands_from_stdin_to_sensors,
-                kwargs={"stop_signal": stop_signal},
-                daemon=True,
-            )
+        try:
+            if self.interactive:
+                interactive_commands_thread = threading.Thread(
+                    name="InteractiveCommandsThread",
+                    target=self._send_commands_from_stdin_to_sensors,
+                    kwargs={"stop_signal": stop_signal},
+                    daemon=True,
+                )
 
-            interactive_commands_thread.start()
+                interactive_commands_thread.start()
 
-        elif self.routine is not None:
-            routine_thread = threading.Thread(
-                name="RoutineCommandsThread",
-                target=self.routine.run,
-                kwargs={"stop_signal": stop_signal},
-                daemon=True,
-            )
-            routine_thread.start()
+            elif self.routine is not None:
+                routine_thread = threading.Thread(
+                    name="RoutineCommandsThread",
+                    target=self.routine.run,
+                    kwargs={"stop_signal": stop_signal},
+                    daemon=True,
+                )
+                routine_thread.start()
+
+            # Wait for the stop signal before exiting.
+            while stop_signal.value == 0:
+                time.sleep(5)
+
+        finally:
+            if not self.stop_sensors_on_exit:
+                return
 
-        # Wait for the stop signal before exiting.
-        while stop_signal.value == 0:
-            time.sleep(5)
+            for command in STOP_COMMANDS:
+                self.serial_port.write(command.encode("utf_8"))
```

Comment:
Also we might need sleep() for like 5sec in between commands. I am not sure, how the hardware will react. Will test it as soon as it is back online